### PR TITLE
Bugfixes in the kernel, added sl_sched_loop_nonblock api, and updated baremetal instructions

### PR DESCRIPTION
**Pull request by phanikishoreg**
### Summary of this Pull Request (PR)

Main change is to incorporate feedback from PR#289.
In addition to that, I've updated the baremetal.md instructions.

### Intent for your PR

Choose one (Mandatory):

- [ ] This PR is for a code-review and is intended to get feedback, but not to be pulled yet.
- [x] This PR is mature, and ready to be integrated into the repo.

### Reviewers (Mandatory):
@gparmer 

### Code Quality

As part of this pull request, I've considered the following:

[Style](https://github.com/gparmer/composite/raw/ppos/doc/style_guide/composite_coding_style.pdf):

- [x] Comments adhere to the Style Guide (SG)
- [x] Spacing adhere's to the SG
- [x] Naming adhere's to the SG
- [x] All other aspects of the SG are adhered to, or exceptions are justified in this pull request
- [-] I have run the auto formatter on my code before submitting this PR (see doc/auto_formatter.md for instructions)

[Code Craftsmanship](http://www2.seas.gwu.edu/~gparmer/posts/2016-03-07-code-craftsmanship.html):

- [x] I've made an attempt to remove all redundant code
- [x] I've considered ways in which my changes might impact existing code, and cleaned it up
- [x] I've formatted the code in an effort to make it easier to read (proper error handling, function use, etc...)
- [x] I've commented appropriately where code is tricky
- [x] I agree that there is no "throw-away" code, and that code in this PR is of high quality

### Testing

I've tested the code using the following test programs (provide list here):

- microbooter, unitschedtest, unit fprr

### CODE REVIEW

File: src/components/lib/sl/sl.c [hunk 1]
```diff
@@ -31,7 +32,8 @@ sl_cs_enter_contention(union sl_cs_intern *csi, union sl_cs_intern *cached, thdc
 		if (!ps_cas(&g->lock.u.v, cached->v, csi->v)) return 1;
 	}
 	/* Switch to the owner of the critical section, with inheritance using our tcap/priority */
-	if ((ret = cos_defswitch(csi->s.owner, t->prio, g->timeout_next, tok))) return ret;
+	if ((ret = cos_defswitch(csi->s.owner, t->prio, csi->s.owner == sl_thd_thdcap(g->sched_thd) ?
```
Comment (gparmer) [inline src/components/lib/sl/sl.c]: this is too clever.  Just do the `s.owner` initialization in a normal conditional.

File: src/components/lib/sl/sl.c [hunk 2]
```diff
@@ -614,7 +617,7 @@ sl_sched_loop(void)
 			 * states of it's child threads) and normal notifications (mainly activations from
 			 * it's parent scheduler).
 			 */
-			pending = cos_sched_rcv(g->sched_rcv, RCV_ALL_PENDING, timeout,
+			pending = cos_sched_rcv(g->sched_rcv, rfl, timeout,
 						&rcvd, &tid, &blocked, &cycles, &thd_timeout);
 			if (!tid) goto pending_events;
```
Comment (gparmer) [inline src/components/lib/sl/sl.c]: likely prefer a `if (tid) {...}` here instead.  The goto feels a little messy

File: src/components/lib/sl/sl.c [hunk 3]
```diff
@@ -614,7 +617,7 @@ sl_sched_loop(void)
...
```
Comment (gparmer) [inline src/components/lib/sl/sl.c]: does the `g->event_head` not need a critical section?

File: src/components/lib/sl/sl.c [hunk 4]
```diff
@@ -670,10 +673,22 @@ sl_sched_loop(void)
...
 			sl_cs_exit();
-		} while (pending);
+		} while (pending > 0);
 
 		if (sl_cs_enter_sched()) continue;
 		/* If switch returns an inconsistency, we retry anyway */
 		sl_cs_exit_schedule_nospin();
 	}
 }
+
```
Comment (gparmer) [inline src/components/lib/sl/sl.c]: I can't figure out the difference, so this really needsa  comment

**Comment by gparmer** [comment]
Since these are bug fixes, I'll accept them now, but some changes/additions are requried.

File: src/components/lib/sl/sl.c (hunk 3 above)
Comment (phanikishoreg) [reply src/components/lib/sl/sl.c]: no it doesn't. this data structure is only accessed  by the scheduler thread in `sl_sched_loop` no where else. i'll remember to add a comment about that in my next PR.

File: src/components/lib/sl/sl.c (hunk 2 above)
Comment (phanikishoreg) [reply src/components/lib/sl/sl.c]: yes, it makes sense. i'll remember to make this change too in my next PR.

File: src/components/lib/sl/sl.c (hunk 1 above)
Comment (phanikishoreg) [reply src/components/lib/sl/sl.c]: I don't understand the comment. You mean initialize `timeout`? Can you please elaborate?

File: src/components/lib/sl/sl.c (hunk 4 above)
Comment (phanikishoreg) [reply src/components/lib/sl/sl.c]: What line/change is this comment referring to?

**Comment by phanikishoreg** [comment]
Thank you for the merge.

**Pull request merged**